### PR TITLE
WIP: Add Azure provider (peer-review PR)

### PR DESCRIPTION
**Pull request by svanharmelen**

### CODE REVIEW

**Comment by radeksimko** [comment]
#2052

File: builtin/providers/azure/resource_azure_disk.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	d.Set("name", datadisk.DiskName)
+	d.Set("lun", datadisk.Lun)
+	d.Set("size", datadisk.LogicalDiskSizeInGB)
+	d.Set("caching", datadisk.HostCaching)
+	d.Set("media_link", datadisk.MediaLink)
+
+	log.Printf("[DEBUG] Retrieving disk: %s", d.Id())
+	disk, err := virtualmachinedisk.NewClient(mc).GetDisk(d.Id())
```
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: This is just a nit but why not create the client once and reuse it? Does not make sense to create a new client for each call, unless something mutates in the client structure after each call.

Something like:

```go
mc := meta.(management.Client)
diskClient, err := virtualmachinedisk.NewClient(mc)
if err != nil {
    return fmt.Errorf("...")
}
disk, err := diskClient.GetDisk(d.Id())
dataDisk, err := diskClient.GetDataDisk(vm, vm, vm, lun)
```

File: builtin/providers/azure/resource_azure_disk.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+	mc := meta.(management.Client)
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	if d.HasChange("size") {
+		p := virtualmachinedisk.UpdateDiskParameters{
+			Name: d.Id(),
+		}
+
+		if d.HasChange("size") {
```
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: You are already in an if block that tests if this condition is true. I think it should be safe to drop this second if, and simply do the:

```go
p.ResizedSizeInGB = d.Get("size").(int)
```

File: builtin/providers/azure/resource_azure_disk.go (hunk 1 above)
Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_disk.go]: Yeah, that would indeed be cleaner... I'm currently adding acceptance tests and will update these (where applicable across all resources) as well.

File: builtin/providers/azure/resource_azure_disk.go (hunk 2 above)
Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_disk.go]: Your currently reviewing the latest resource that was added and which is not fully finished/tested yet, so it's expected to have a few of these...

But thanks for pointing it out as this is a leftover from some refactored bits that indeed needs to be changed/updated.

Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: Not to worry. Just doing a quick look over the code. More comments will come as I go through it. I will then leave a comment on the PR itself if needed. Peer review is meant to be helpful in making code and/or the general architecture better.

Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_disk.go]: I know, just letting you know :wink:

File: builtin/providers/azure/resource_azure_disk.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+	if err != nil {
+		return fmt.Errorf("Error retrieving disk %s: %s", d.Id(), err)
+	}
+
+	d.Set("virtual_machine", disk.AttachedTo.RoleName)
+
+	return nil
+}
+
+func resourceAzureDataDiskUpdate(d *schema.ResourceData, meta interface{}) error {
```
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: This function seams to be a bit long. I wonder if its not better to abstract away some of the boilerplate that happens during most calls to the API. Looks like every call is 3 step:

* compose payload
* call API
* wait for result

A small wrapper around the azure API might help us de-duplicate some code and also make things easier to test.

So for example, we could have a type for virtual disk management that exposes some high level operations. And in this function we would do something similar to (pseudo code ahead):

```go
for {
    switch{
    case d.HasChange("size"):
        err := vmdisk.UpdateDisk(d)
    case d.HasChange("virtual_machine"):
        err := vmdisk.SetVirtualMachine(d)
    case d.HasChange("caching"), d.HasChange("lun"):
        err := vmdisk.UpdateDataDisk(d)
    default:
        return resourceAzureDataDiskRead(d, meta)
    }
}
```
This could be done for all other resources that duplicate code. These functions seem to be entry points for terraform. I believe it is best to remove as much of the business code from them into a separate type/interface/package as possible.  This way you decouple the implementation from the entry points, allowing you to easily switch it, if you need to, in the future. It also makes it easier for you to test and extend the functionality, without piling everything on top of a single function.

Thoughts?

File: builtin/providers/azure/resource_azure_disk.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	d.SetId("")
+
+	return nil
+}
+
+func hostCaching(d *schema.ResourceData) virtualmachinedisk.HostCachingType {
+	switch d.Get("caching").(string) {
+	case "ReadOnly":
```
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: hmm. I don't think this is necessary. You should be able to cast directly to ```HostCachingType```, and you would get one of:
```go
HostCachingTypeNone     
HostCachingTypeReadOnly 
HostCachingTypeReadWrite
```

File: builtin/providers/azure/resource_azure_disk.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+	if !ok {
+		name = fmt.Sprintf("%s-%d", d.Get("virtual_machine").(string), d.Get("lun").(int))
+	}
+
+	return fmt.Sprintf(dataDiskBlobStorageURL, d.Get("storage").(string), name.(string))
+}
+
+func verifyDataDiskParameters(d *schema.ResourceData) error {
+	caching := d.Get("caching").(string)
+	if caching != "None" && caching != "ReadOnly" && caching != "ReadWrite" {
```
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: If you cast to ```HostCachingType``` you should be able to do a simple switch here:

switch caching{
case HostCachingTypeNone, HostCachingTypeReadOnly, HostCachingTypeReadWrite:
default:
    return fmt.Errorf()
}

I generally avoid using magic strings. Using constants and variables avoids the need to do a search and replace if the authors of the 3rd party module decide to change those strings for whatever reason.

**Comment by gabriel-samfira** [comment]
Will continue review later. So far just a few comments about 1/3 of the way in.

File: builtin/providers/azure/resource_azure_instance.go [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+				ForceNew: true,
+			},
+
+			"username": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"password": &schema.Schema{
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: Same as above.

File: builtin/providers/azure/resource_azure_instance.go [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+	cs, err := hostedservice.NewClient(mc).GetHostedService(d.Id())
+	if err != nil {
+		return fmt.Errorf("Error retrieving Cloud Service of instance %s: %s", d.Id(), err)
+	}
+
+	d.Set("reverse_dns", cs.ReverseDNSFqdn)
+	d.Set("location", cs.Location)
+
+	log.Printf("[DEBUG] Retrieving instance: %s", d.Id())
+	dpmt, err := virtualmachine.NewClient(mc).GetDeployment(d.Id(), d.Id())
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: Separate hosted service name here.

File: builtin/providers/azure/resource_azure_instance.go [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+
+	log.Printf("[DEBUG] Creating Cloud Service for instance: %s", name)
+	err = hostedservice.NewClient(mc).CreateHostedService(p)
+	if err != nil {
+		return fmt.Errorf("Error creating Cloud Service for instance %s: %s", name, err)
+	}
+
+	// Put in this defer here, so we are sure to cleanup already created parts
+	// when we exit with an error
+	defer func(mc management.Client) {
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: Wasn't Terraform supposed to not mess with anything if a failure had occurred to facilitate someone coming in and picking off where it had stopped?
If so, the existence of this defer statement is questionable.

In the case of separate hosted service resources; this would have not been necessary anyhow, because they would have had to have been created beforehand...

File: builtin/providers/azure/resource_azure_instance.go [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+		return err
+	}
+
+	// Verify if we have all required parameters
+	if err := verifyInstanceParameters(d, osType); err != nil {
+		return err
+	}
+
+	p := hostedservice.CreateHostedServiceParameters{
+		ServiceName:    name,
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: Considering hosted services to be a separate entity; checking for the presence of an optional hosted_service parameter and using it here would be required, else default to the name of the instance if no such parameter was provided.

File: builtin/providers/azure/resource_azure_instance.go (hunk 6 above)
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: Same as above with username.

File: builtin/providers/azure/resource_azure_instance.go [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+				ForceNew: true,
+			},
+
+			"time_zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"username": &schema.Schema{
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: These should not be mandatory due to chains of events like having a VM configured to join an Active Directory domain etc...

File: builtin/providers/azure/resource_azure_disk.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+	case "ReadOnly":
+		return virtualmachinedisk.HostCachingTypeReadOnly
+	case "ReadWrite":
+		return virtualmachinedisk.HostCachingTypeReadWrite
+	default:
+		return virtualmachinedisk.HostCachingTypeNone
+	}
+}
+
+func mediaLink(d *schema.ResourceData) string {
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: See above comment regarding this function and dataDiskBlobStorageURL

File: builtin/providers/azure/resource_azure_disk.go [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+
+	return resourceAzureDataDiskRead(d, meta)
+}
+
+func resourceAzureDataDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	// If a name was not supplied, it means we created a new emtpy disk and we now want to
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: Your logic here is pretty hard to follow; the feature of an actively chosen delete/detach option is a great idea though.
Personally; I would make a separate parameter called ephemeral or whatever and judge the disk's deletion/detachment according to that one.

File: builtin/providers/azure/resource_azure_disk.go [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+			"Error waiting for data disk %d to be added to instance %s: %s", lun, vm, err)
+	}
+
+	log.Printf("[DEBUG] Retrieving data disk %d from instance %s", lun, vm)
+	disk, err := virtualmachinedisk.NewClient(mc).GetDataDisk(vm, vm, vm, lun)
+	if err != nil {
+		return fmt.Errorf("Error retrieving data disk %d from instance %s: %s", lun, vm, err)
+	}
+
+	d.SetId(disk.DiskName)
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: Using the name as ID is technically sound as the way Azure's API is currently implemented makes name collisions unacceptable; but why not use the label fields every Azure entity offers for both internal and remote tracking and just use the resource name for the humans looking at the dashboard?

File: builtin/providers/azure/resource_azure_disk.go [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+		return fmt.Errorf("Error adding data disk %d to instance %s: %s", lun, vm, err)
+	}
+
+	// Wait until the data disk is added
+	if err := mc.WaitForOperation(req, nil); err != nil {
+		return fmt.Errorf(
+			"Error waiting for data disk %d to be added to instance %s: %s", lun, vm, err)
+	}
+
+	log.Printf("[DEBUG] Retrieving data disk %d from instance %s", lun, vm)
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: Avoiding using names themselves for tracking (see below comment) could dodge the re-read of the freshly created resource entirely, especially considering a Read is done at the end of Create as well.

File: builtin/providers/azure/resource_azure_disk.go [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+		MediaLink:           mediaLink(d),
+		SourceMediaLink:     d.Get("source_media_link").(string),
+	}
+
+	if name, ok := d.GetOk("name"); ok {
+		p.DiskName = name.(string)
+	}
+
+	log.Printf("[DEBUG] Adding data disk %d to instance: %s", lun, vm)
+	req, err := virtualmachinedisk.NewClient(mc).AddDataDisk(vm, vm, vm, p)
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: This is troublesome when you consider instances and hosted services as decoupled and independent resources.
I'd go ahead and add an optional service_name parameter and use that if it's given; else default to the vm's name.

File: builtin/providers/azure/resource_azure_disk.go [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+
+	p := virtualmachinedisk.CreateDataDiskParameters{
+		Lun:                 lun,
+		LogicalDiskSizeInGB: d.Get("size").(int),
+		HostCaching:         hostCaching(d),
+		MediaLink:           mediaLink(d),
+		SourceMediaLink:     d.Get("source_media_link").(string),
+	}
+
+	if name, ok := d.GetOk("name"); ok {
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: You could make names mandatory; see below comments.

File: builtin/providers/azure/resource_azure_disk.go [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/svanharmelen/azure-sdk-for-go/management"
+	"github.com/svanharmelen/azure-sdk-for-go/management/virtualmachinedisk"
+)
+
+const dataDiskBlobStorageURL = "http://%s.blob.core.windows.net/disks/%s.vhd"
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_disk.go]: This should preferably be exported in the sdk or the below function which uses it below be moved to the sdk.

File: builtin/providers/azure/provider.go [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+// Provider returns a terraform.ResourceProvider.
+func Provider() terraform.ResourceProvider {
+	return &schema.Provider{
+		Schema: map[string]*schema.Schema{
+			"settings_file": &schema.Schema{
```
Comment (aznashwan) [inline builtin/providers/azure/provider.go]: Would have been nice to provide the id + certificate + url authentication option as well for people with Windows Azure Packs.

File: builtin/providers/azure/resource_azure_instance.go [hunk 19]
```diff
@@ -0,0 +1,653 @@
...
+		if img.Label == label {
+			if img.OS != linux && img.OS != windows {
+				return nil, "", fmt.Errorf("Unsupported image OS: %s", img.OS)
+			}
+			if img.MediaLink == "" {
+				if storage == "" {
+					return nil, "",
+						fmt.Errorf("When using a platform image, the 'storage' parameter is required")
+				}
+				img.MediaLink = fmt.Sprintf(osDiskBlobStorageURL, storage, label)
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: osDiskBlobStorageURL should ideally be moved into the sdk. See above.

File: builtin/providers/azure/resource_azure_instance.go [hunk 20]
```diff
@@ -0,0 +1,653 @@
...
+		}
+
+		if _, ok := d.GetOk("time_zone"); !ok {
+			return fmt.Errorf("You must supply a 'time_zone' when using a Windows image")
+		}
+	}
+
+	if _, ok := d.GetOk("subnet"); ok {
+		if _, ok := d.GetOk("virtual_network"); !ok {
+			return fmt.Errorf("You must also supply a 'virtual_network' when supplying a 'subnet'")
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_instance.go]: These kinds of parameter dependencies checked directly through the schema would be awesome.
Until then, good job enforcing them here!

File: builtin/providers/azure/resource_azure_disk.go (hunk 13 above)
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_disk.go]: Does the azure API not offer a unique ID per resource? I find it a bit...miopic if they do not.

File: builtin/providers/azure/resource_azure_instance.go (hunk 10 above)
Comment (gabriel-samfira) [inline builtin/providers/azure/resource_azure_instance.go]: To join an active directory domain, you need a local administrator and the credentials to an AD administrator. So in theory, when joining an AD domain, you still need to either have a local admin whoms credentials you know, or create a new one when joining a domain.

File: builtin/providers/azure/resource_azure_virtual_network.go [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+		},
+	}
+}
+
+func resourceAzureVirtualNetworkCreate(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(management.Client)
+
+	name := d.Get("name").(string)
+
+	nc, err := virtualnetwork.NewClient(mc).GetVirtualNetworkConfiguration()
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_virtual_network.go]: I must insist on the fact that networking operations need to be synced somehow.
Long story short of what's being presented [here](https://msdn.microsoft.com/en-us/library/azure/jj157182.aspx), because of how the virtualnetwork API of Azure works (i.e. you ask for all the networking configuration associated to your account, do the necessary additions/modifications/removals on it, and send it all back to Azure where it effectively gets 'pasted over' the current resources in Azure), it is extremely hazardous to do such operations currently.

With your current implementation; what would happen is that, whenever two or more virtual networks (the only resource currently affected by this problem here) are defined and apply is ran; they'd go ahead, each would fetch its own network configuration and 'apply itself' unto it; after which both would send the config back (both networks not having any idea about each-other's creation), and the one that sends the config last practically overrides the first comer's creation.

If you don't believe me; please attempt creating two or more virtual networks in one go. You'll see that only one of them actually gets created; the others fading silently away from existence because you call Read at the end of Create and Terraform assumes the resource has been deleted in the meantime...

By far the most feasible solution here is simply using a mutex whenever resources affected by this limitation of the API  (virtual networks here; local network connections and DNS definitions on my side as well) are Created/Read/Deleted.
Sure, it sucks big time; as the updating of such resources would have to be sequentially. One must bear in mind, however, that this is wholly a limitation of the Azure API and there's really nothing us as Terraform developers can do that hit the +1 button [here](http://feedback.azure.com/forums/217313-networking-dns-traffic-manager-vpn-vnet/suggestions/5954992-make-virtual-network-operations-atomic-on-portal-a). :disappointed:

Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_virtual_network.go]: I fully understand how the API works with regards to virtual networks, and indeed the API for virtual networks cannot be used concurrently in a safe way.

So yes, you are completely right that these network operation need to be synced somehow. But let me explain why your suggestion to fix it in the Azure provider itself isn't going to work...

In your implementation you added a mutex to guard concurrent use of a `management.Client` within the plugin, to prevent these kinds of problems with API endpoints that do not support being used concurrently. But there is no concurrency within the plugin itself. The concurrency is build into the core of Terraform and it will create a new instance of the plugin (most likely in it's own goroutine) for every call that can be made concurrently.

So in this case if you create a Terraform config with two virtual networks in there, Terraform core will actually call the plugin twice, resulting in two separate instances of the plugin that will then both have their own mutex.

Considering all this, you'll now probably understand that having a mutex here (or any other means to synchronise access to the `management.Client` within the plugin itself) will not solve your problem as it will only guard an already synchronous instance of the plugin from using the `management.Client` concurrently.

So my line of thought here was that we currently have only one, or maybe two (both not feasible), options to fix this:

1. Fix the Azure API (the only real solution if you ask me);
2. Change the internals of Terraform;

Unfortunately these are both changes that most likely will not be made anytime (if ever) soon. Next to that our influence of getting it changed it is very limited at the least. So that leaves us with only one additional option which would be to not offer a resource to manage virtual networks. But that in turn would be taking away a significant portion of the functionality and with that the usability of the Azure provider...

So I decided (for myself) that for the time being the best approach would be to just build the provider/resources as if it was safe to make concurrent API calls and then document very clearly in the docs that for the `azure_virtual_network` resource there currently is a known limitation/issue with the API.

Luckily we can offer a (pretty simple) workaround in those same docs, so people can safely use Terraform configs that contain multiple `azure_virtual_network` resources. The workaround is to add an explicit dependency between all `azure_virtual_network` resources in your config by defining the `depends_on` argument for these resources. That way Terraform will make all calls to create virtual network synchronously, one after the other...

Is it perfect? Nope! Is it an acceptable way forward at this very moment in time? Yes I believe so...

Comment (aznashwan) [inline builtin/providers/azure/resource_azure_virtual_network.go]: I do admit; I did not take as thorough a look through the internals of Terraform as I would have liked.

The exact way of handling provider plugins you're describing was my exact fear when I first hit this problem.

Weirdly enough though; I can personally vouch that using a mutex there __does__ take care of the problem __by itself__ (i.e. I did not introduce any explicit config dependencies in my attempts; just defined two independent virtual networks and let Terraform do its thing...).

I feel it would be best to ask someone more knowledgeable in Terraform's internals before we continue this debate further.
Will come back to you on this as soon as I get my clarifications...

Comment (jen20) [inline builtin/providers/azure/resource_azure_virtual_network.go]: I implemented a (still closed-source provider since not had permission to open source it yet...) terraform provider for Azure that ran into this exact issue. The only way to get around it was to serialise the access to the API as you say. However, I split the virtual network resource into network and subnet resources so the problem was more acute - the graph resolution would be such that with no explicit dependency management that both subnets could be created at once, and would do so in different provider processes - in the end a mutex file was required and one or the other had to block even though terraform thought it was constructing them in parallel. Sadly this still doesn't stop concurrency issues from other management clients elsewhere. The API is just broken and it seems that Microsoft don't care enough to fix it.

It's also worth discussing with the azure-sdk-for-go guys their plans to switch over to an auto-generated API based on the Resource Manager API instead of the Service Management API - it might end up rendering a big chunk of this obsolete in reasonably short order.

Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_virtual_network.go]: Well sir, I stand corrected!

After repeating your tests with a customised branch of the one in this PR, I could indeed clearly see that your absolutely right!

Looking a little deeper into the code it looks like TF is reusing existing connections to already connected plugins by multiplexing over those existing connections (see [these](https://github.com/hashicorp/terraform/tree/master/rpc) files for more details). 

Luckily it's just a small change to add the mutex so I updated the code accordingly just before I finished writing the docs. So the Virtual Network related API calls will now by synchronised :smiley:

As in this case, it often seems that direct communication (like over IRC in this case) works a lot better to discuss stuff like this and/or to come to a resolution. So please know that you can almost always find me on IRC if anything is up.

Thanks for pointing this out! Will push in the updated code and docs in a few minutes as soon as a last round of acceptance tests are finished (ETA 15 min, they take about 40 minutes)...

Comment (aznashwan) [inline builtin/providers/azure/resource_azure_virtual_network.go]: Glad to have helped.
When I tested it; I wasn't so thorough as looking through the code as I was of just running it and seeing what happens, so I got to learn some more about Terraform's internals today, which is nice.

Good job on everything and god*speed* merging!
Can't wait to see your work get in and start expanding it even further from there on :smile:

File: builtin/providers/azure/provider.go [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+		ConfigureFunc: providerConfigure,
+	}
+}
+
+func providerConfigure(d *schema.ResourceData) (interface{}, error) {
+	config := Config{
+		SettingsFile:   d.Get("settings_file").(string),
+		SubscriptionID: d.Get("subscription_id").(string),
+		Certificate:    []byte(d.Get("certificate").(string)),
+	}
```
Comment (phinze) [inline builtin/providers/azure/provider.go]: Can you do a homedir expand on the `settings_file`? Like we do [over here](https://github.com/hashicorp/terraform/blob/434d0c9d86c998dbb229c043467d80097e2d85c6/config/interpolate_funcs.go#L64).

File: builtin/providers/azure/resource_azure_data_disk.go [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+}
+
+func resourceAzureDataDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	mc := meta.(*Client).mgmtClient
+
+	lun := d.Get("lun").(int)
+	vm := d.Get("virtual_machine").(string)
+
+	// If a name was not supplied, it means we created a new emtpy disk and we now want to
+	// delete that disk again. Otherwise we only want to detach the disk and keep the blob.
```
Comment (phinze) [inline builtin/providers/azure/resource_azure_data_disk.go]: This "sometimes a disk, sometimes just an attachment" modeling seems a bit muddy to me, but it seems to match the upstream's API, so this seems good for now.

File: builtin/providers/azure/resource_azure_instance.go (hunk 8 above)
Comment (phinze) [inline builtin/providers/azure/resource_azure_instance.go]: These are good questions. Going to land this as-is and we'll circle back on the discussion of separating hosted service into its own resource.

**Comment by phinze** [comment]
Okay nearly completed my review pass. I've kicked off the AccTests locally and I have to run to a meeting. Will complete review afterwards and we can get this in. :ok_hand:

**Comment by svanharmelen** [comment]
I'll be behind a desk again later today, so will have a look in a few hours...

**Comment by phinze** [comment]
All AccTest passed locally. This is looking good to me. My nits can be cleaned up post merge if you want - @svanharmelen - but will let you decide what you want to do. From my side: merge at will!

File: builtin/providers/azure/provider.go (hunk 22 above)
Comment (svanharmelen) [inline builtin/providers/azure/provider.go]: Updated, will commit in a second...

File: builtin/providers/azure/resource_azure_data_disk.go (hunk 23 above)
Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_data_disk.go]: Yeah, I understand what you mean but indeed this follows the API... Maybe we can/should add some more attributes to make it more conscious choices/behaviour?

But I also think it's good for now, so I'll leave this as is for the time being.

**Comment by svanharmelen** [comment]
Fixed/updated the retrieval of the `settings_file` path so the homedir (if given with the shorthand `~`) is properly expanded to a full path. Travis seems to like everything and is green, so merging it in :smiley:

**Comment by phinze** [comment]
:tada: :confetti_ball:

File: builtin/providers/azure/resource_azure_virtual_network_test.go [hunk 24]
```diff
@@ -0,0 +1,295 @@
...
+
+		if network.Location != "West US" {
+			return fmt.Errorf("Bad location: %s", network.Location)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckAzureVirtualNetworkDestroy(s *terraform.State) error {
```
Comment (aznashwan) [inline builtin/providers/azure/resource_azure_virtual_network_test.go]: Shouldn't this function just check if the resource still exists instead of actually deleting it itself?

Comment (phinze) [inline builtin/providers/azure/resource_azure_virtual_network_test.go]: Yep that's correct. Good catch!

Comment (svanharmelen) [inline builtin/providers/azure/resource_azure_virtual_network_test.go]: Indeed a good catch! I think I might have used this (wrong) approach in the CloudStack provider as well. Will walk through both providers and update the tests where applicable...
